ambry-api/src/main/java/com/github/ambry/rest/RestUtils.java

```diff
@@ -860,38 +860,50 @@ public static void accountAndContainerNamePreconditionCheck(RestRequest restRequ
 
   /**
    * Sets the user metadata in the headers of the response.
-   * @param userMetadata the user metadata that needs to be sent.
+   * @param userMetadata byte array containing user metadata that needs to be sent.
    * @param restResponseChannel the {@link RestResponseChannel} that is used for sending the response.
    * @return {@code true} if the user metadata was successfully deserialized into headers, {@code false} if not.
    * @throws RestServiceException if there are any problems setting the header.
    */
-  public static boolean setUserMetadataHeaders(byte[] userMetadata, RestResponseChannel restResponseChannel)
-      throws RestServiceException {
+  public static boolean setUserMetadataHeaders(byte[] userMetadata, RestResponseChannel restResponseChannel) {
     Map<String, String> userMetadataMap = buildUserMetadata(userMetadata);
-    boolean setHeaders = userMetadataMap != null;
-    if (setHeaders) {
-      for (Map.Entry<String, String> entry : userMetadataMap.entrySet()) {
-        String headerName = entry.getKey();
-        String headerValue = entry.getValue();
+    if (userMetadataMap != null) {
+      setUserMetadataHeaders(userMetadataMap, restResponseChannel);
+      return true;
+    } else {
+      return false;
+    }
+  }
+
+  /**
+   * Sets the user metadata in the headers of the response.
+   * @param userMetadataMap map of user metadata that needs to be sent.
+   * @param restResponseChannel the {@link RestResponseChannel} that is used for sending the response.
+   * @throws RestServiceException if there are any problems setting the header.
+   */
+  public static void setUserMetadataHeaders(Map<String, String> userMetadataMap,
```

Comment:
There seems to be a difference between the map provided in open source and the closed source place where this is intended to be used.

`RestUtils.buildUserMetadata` returns a map with keys that already have the "x-ambry-um-" prefix.
The decoded map in closed source does not (without writing some code that adds them).

Would it be possible to make `buildUserMetadata` not prepend the prefix and have `setUserMetadataHeaders` do it instead?

Reply:
Agree. Also I am trying to understand the motivation to have this new method. Are you planning to replace the piece of similar logic after `codec.decode` in closed source to avoid duplicate code?

Reply:
Nvm, you have explained offline.

Reply:
I updated the new method to add the prefix if not already present.